runtime/calamari/src/migrations/staking.rs

```diff
@@ -0,0 +1,163 @@
+        From<<T as frame_system::Config>::AccountId>,
+    pallet_parachain_staking::BalanceOf<T>: Into<Balance> + From<Balance>,
+{
+    fn on_runtime_upgrade() -> Weight
+    where
+        <<T as frame_system::Config>::Origin as OriginTrait>::AccountId:
+            From<<T as frame_system::Config>::AccountId>,
+        pallet_parachain_staking::BalanceOf<T>: Into<Balance> + From<Balance>,
+    {
+        log::info!(
+            target: "SplitDelegatorStateIntoDelegationScheduledRequests",
+            "running migration for DelegatorState to new version and DelegationScheduledRequests \
+            storage item"
+        );
+
+        // let mut reads: Weight = 0;
+        // let mut writes: Weight = 0;
+
+        // 1. Find current invulnerables
+        let invulnerables = manta_collator_selection::Pallet::<T>::invulnerables();
+
+        // 2. Clear the invulnerables list
+        let _ = manta_collator_selection::Pallet::<T>::set_invulnerables(
+            <T as frame_system::Config>::Origin::root(),
+            Vec::new(),
+        );
+
+        // 3. Register invulnerables to whitelist
+        // i.e. onboard with manta_collator_selection::registerCandidate
+        for invuln in invulnerables.clone() {
+            let _ = manta_collator_selection::Pallet::<T>::register_candidate(
+                <T as frame_system::Config>::Origin::root(),
+                invuln,
+            );
+        }
+
+        // 4. Initialize parachain staking pallet to genesis-equivalent state
+        //    and onboard invulnerables to staking with 400k lock
+
+        // +1 because migrations execute before this is updated in the on_initialize hook
+        let current_block = frame_system::Pallet::<T>::block_number() + 1u32.into();
+        pallet_parachain_staking::Pallet::<T>::initialize_pallet(
+            current_block,
+            invulnerables,
+            crate::currency::inflation_config::<T>(),
+        );
+        // Setting total_selected will take effect at the beginning of the next round, so for the first 6 hours
+        // our invulnerables will be the only collators
+        let _ = pallet_parachain_staking::Pallet::<T>::set_total_selected(
+            <T as frame_system::Config>::Origin::root(),
+            63u32,
+        );
+
+        // TODO: Get correct weight from the extrinsics
+        // T::DbWeight::get().reads_writes(reads, writes)
+        T::BlockWeights::get().max_block // only migration, we can just use the whole block
+    }
+
+    #[cfg(feature = "try-runtime")]
+    fn pre_upgrade() -> Result<(), &'static str> {
+        use crate::sp_api_hidden_includes_construct_runtime::hidden_include::traits::Currency;
+
+        // 2. Ensure they each have 400k KMA
+        let invulnerables = manta_collator_selection::Pallet::<T>::invulnerables();
+        for invulnerable in invulnerables.clone() {
+            assert!(
+                <T as pallet_parachain_staking::Config>::Currency::free_balance(&invulnerable)
```

Comment:
So they need to have 400k extra free or is their old locked balance unlocked at this point ?

Reply:
join_candidates unlocks their balance if it's in the manta_coll_sell::candidates() storage.

But even if it didn't: Locks are not reservations. Multiple locks overlap. So the first 400k would be locked twice, it wouldn't lock 800k total

Reply:
The invulnerables must have 400k+ in total free balance available before the upgrade.

collator selection *reserves* 400k for a candidate and we add them to it because that's the whitelist
This 400k is then *unreserved* in join_candidate and the free balance created from it is then *locked* for the candidate deposit